internal/bzlmod/go_deps.bzl

```diff
@@ -15,6 +15,26 @@ def _repo_name(importpath):
     candidate_name = "_".join(segments).replace("-", "_")
     return "".join([c.lower() if c.isalnum() else "_" for c in candidate_name.elems()])
 
+_GO_DEPS_HELPER_DEFS_BZL = """load("@bazel_gazelle//internal/bzlmod:go_helper.bzl", "go_helper")
+
+def go(import_path):
```

Comment:
Personally I don't view the name `go` as an affront. If someone finds themselves confused by it they can alias it to be something different. 

That said, I'm slightly concerned about having a macro that transmogrifies the import path `github.com/bazelbuild/rules_go/foo` into the bzlmod form. This would mean that users are unable to grep their codebase for the string that actually shows up in their BUILD file.

Why have we decided to go this route now? It seems we were able to make the transition to bzlmod so this can't be a precondition. Is it?

If we don't think the bazel syntax is good enough, maybe we should turn `go_library` into a macro and have it invoke the `go` function when it detected something that isn't label shaped?

Reply:
Sorry, we had a meeting where we discussed some of the context for this, I'm happy to explain the motivation.

> It seems we were able to make the transition to bzlmod so this can't be a precondition. Is it?

The transition to bzlmod is not really complete, because it currently requires users to add every third-party go repository to the `use_repo` call in their `MODULE.bazel` file.
https://github.com/bazelbuild/bazel-gazelle/pull/1404 is one proposed solution to this, which adds another code-generation mechanism to update this section of the file. However, it's making Go unique in the bazel ecosystem, which I think makes it harder for users.

Every other language that works well under bzlmod has a form where users only load from a single "trampoline" repo for the whole transitive closure they imported:

- Java: https://github.com/bazelbuild/rules_jvm_external/blob/master/examples/bzlmod/java/src/com/github/rules_jvm_external/examples/bzlmod/BUILD#L7
- JavaScript: https://github.com/aspect-build/rules_js/blob/main/e2e/bzlmod/BUILD.bazel#L4
- Python: https://github.com/bazelbuild/rules_python/blob/main/examples/bzlmod/BUILD.bazel#L18

This became really evident as I have started delivering a Bazel training course which is exclusively bzlmod. https://github.com/aspect-build/codelabs/tree/final is the final state of the codelab which illustrates how languages are used "kinda the same" but Go is the outlier right now, as the `use_repo` calls in the MODULE.bazel file illustrate.

> grep their codebase for the string that actually shows up in their BUILD file
doesn't this new option improve the situation, now that the string in the BUILD file is the same as the one in the `import` statement in the `.go` file?

There's a whole separate discussion of whether we'd want Gazelle to support this form, but I think as written this PR is just adding a syntax sugar convenience and not tying us to any particular way forward.

Reply:
Sorry @achew22 for not providing the context right away, but @alexeagle did a great job explaining it.

Let me add that I can see #1404 being the right approach for the "do everything in Bazel camp" and this PR to please the "stay as close as possible to the native tools" demographic. They each have their own pros and cons and thus may have (but also can) coexist. 

One nice feature I'm currently implementing for the `go` macro is strict deps: If you depend on a Go package that is missing from your own module's `go.mod` but contained in some Bazel dependency's `go.mod`, you would see a descriptive error rather than risking having your build broken by changes in dependencies.